### ISSUE TITLE
Write kernels like process_write_paged() occasionally read past end of dispatch CB giving data mismatch

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent-metal/tt-metal#7061]
While working on new unit tests for new FD2 write kernels in dispatcher (more loops), a mismatch was seen where instead of data written to memory, zeros were written.    Upon closer look I see that it happens due to `cb_fence` and then `data_ptr` exceeding L1 1.5MB size (`worker_l1_size: 1499136 bytes`), and not correctly wrapping back to base of dispatch CB buffer `dispatch_cb_base` via `cb_fence` logic. Not certain how it's supposed to work yet with prefetch/dispatch blocking scheme, studying closer.. probably a small bug somewhere.

May affect other write variants (linear write) but observed in paged-write here.

Repro at any recent fd-2/main commit, including most recent 5e825d804 from Apr2.

Trimed Failing test (write of 128x 8KB pages to DRAM, multiplied by 2 loops in spoof_prefetcher.cpp kernel):

```
TT_METAL_SLOW_DISPATCH_MODE=1 ./build/test/tt_metal/perf_microbenchmark/dispatch/test_dispatcher -w 0 -t 2 -wx 0 -wy 1 -min 8192 -max 8192 -lps 13 -pbs 2 -np 128 -c -i 1 -pi 2 |& tee fail.log

<snip>

Doing loop iter with xfer_size: 8192 dst: 1c0040086000 data_ptr: 16c010 cb_fence: 19c000 <=== cb_fence too high
Writing xfer_size: 8192 to addr: 40086000 noc_xy: 1c00 write_length: 729088 data_ptr: 16c010 <== Safe data_ptr
data[0]: 27000000
data[1]: 27000001
data[2]: 27000002
data[3]: 27000003
data[4]: 27000004
data[5]: 27000005
data[6]: 27000006
data[7]: 27000007
Yes done page_id: 40 for page_size: 8192
Doing loop iter with xfer_size: 8192 dst: 45000086000 data_ptr: 16e010 cb_fence: 19c000 <=== data_ptr exceeds l1_worker_size
Writing xfer_size: 8192 to addr: 86000 noc_xy: 450 write_length: 720896 data_ptr: 16e010 
data[0]: 0
data[1]: 0
data[2]: 0
data[3]: 0
data[4]: 0
data[5]: 0
data[6]: 0
data[7]: 0
Yes done page_id: 41 for page_size: 8192
<snip>

                   Test | FATAL    | Data mismatch - First 20 failures for logical_core: (x=0,y=11) (physical: (x=0,y=11))
                   Test | FATAL    | [8192] (Fail) Expected: 0x30000000 Observed: 0x00000000
                   Test | FATAL    | [8193] (Fail) Expected: 0x30000001 Observed: 0x00000000
                   Test | FATAL    | [8194] (Fail) Expected: 0x30000002 Observed: 0x00000000

```

Passing test (single loop in prefetcher):

```
TT_METAL_SLOW_DISPATCH_MODE=1 ./build/test/tt_metal/perf_microbenchmark/dispatch/test_dispatcher -w 0 -t 2 -wx 0 -wy 1 -min 8192 -max 8192 -lps 13 -pbs 2 -np 128 -c -i 1 -pi 1 |& tee pass.log
```

Debug/fix testing continues today.

**Comment by kmabeeTT** [comment]
Debugged more, I see this is test issue.. dispatch CB buffer is programmed "too big" and overflows L1.  Observed from some initialization prints I added.   

```
Initialized block_next_start_addr for block 0 using offset: 393216 and dispatch_cb_base: 114688 to 0x7c000 (507904)
Initialized block_next_start_addr for block 1 using offset: 786432 and dispatch_cb_base: 114688 to 0xdc000 (901120)
Initialized block_next_start_addr for block 2 using offset: 1179648 and dispatch_cb_base: 114688 to 0x13c000 (1294336)
Initialized block_next_start_addr for block 3 using offset: 1572864 and dispatch_cb_base: 114688 to 0x19c000 (1687552) <== Overflow, no good.
```

Demoting to P1 since not kernel issue, will work on test fix and assert.

**Comment by kmabeeTT** [comment]
Merged test fix and new asserts at 2d9787972, closing.
